pkg/trafficrouting/network/custom/custom.go

```diff
@@ -0,0 +1,306 @@
+			klog.Errorf("failed to get lua script for custom network provider %s: %s", ref.Kind, err.Error())
+			return err
+		}
+		if err := r.storeObject(obj); err != nil {
+			klog.Errorf("failed to store custom network provider %s/%s", ref.Kind, ref.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *customController) EnsureRoutes(ctx context.Context, strategy *rolloutv1alpha1.TrafficRoutingStrategy) (bool, error) {
+	var err error
+	var done = true
+	for _, ref := range r.conf.TrafficConf {
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion(ref.APIVersion)
+		obj.SetKind(ref.Kind)
+		if err = r.Get(ctx, types.NamespacedName{Namespace: r.conf.RolloutNs, Name: ref.Name}, obj); err != nil {
+			return false, err
+		}
+		specStr := obj.GetAnnotations()[OriginalSpecAnnotation]
+		if specStr == "" {
+			continue
+		}
+		var oSpec Data
+		_ = json.Unmarshal([]byte(specStr), &oSpec)
+		luaScript, err := r.getLuaScript(ctx, ref)
+		if err != nil {
+			klog.Errorf("failed to get lua script for %s", ref.Kind)
+			return false, err
+		}
+		nSpec, err := r.executeLuaForCanary(oSpec, strategy, luaScript)
+		if err != nil {
+			return false, err
+		}
+		if cmpAndSetObject(nSpec, obj) {
+			continue
+		}
+		if err = r.Update(context.TODO(), obj); err != nil {
+			klog.Errorf("failed to update custom network provider")
+			return false, err
+		}
+		klog.Infof("update custom network provider %s/%s success")
+		done = false
+	}
+	return done, nil
+}
+
+func (r *customController) Finalise(ctx context.Context) error {
+	for _, ref := range r.conf.TrafficConf {
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion(ref.APIVersion)
+		obj.SetKind(ref.Kind)
+		if err := r.Get(ctx, types.NamespacedName{Namespace: r.conf.RolloutNs, Name: ref.Name}, obj); err != nil {
+			if errors.IsNotFound(err) {
+				klog.Infof("custom network provider %s/%s not found when finalising", ref.Kind, ref.Name)
+				continue
+			}
+			return err
+		}
+		if err := r.restoreObject(obj); err != nil {
+			klog.Errorf("failed to restore object: %s/%s", ref.Kind, ref.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+// store spec of an object in OriginalSpecAnnotation
+func (r *customController) storeObject(obj *unstructured.Unstructured) error {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	labels := obj.GetLabels()
+	oSpec := annotations[OriginalSpecAnnotation]
+	delete(annotations, OriginalSpecAnnotation)
+	data := Data{
+		Spec:        obj.Object["spec"],
+		Labels:      labels,
+		Annotations: annotations,
+	}
+	cSpec := util.DumpJSON(data)
+	if oSpec == cSpec {
+		return nil
+	}
+	annotations[OriginalSpecAnnotation] = cSpec
+	obj.SetAnnotations(annotations)
+	if err := r.Update(context.TODO(), obj); err != nil {
+		klog.Errorf("failed to store custom network provider %s/%s", obj.GetKind(), obj.GetName())
+		return err
+	}
+	klog.Infof("store custom network provider %s/%s success", obj.GetKind(), obj.GetName())
```

Comment:
store custom network provider(%s/%s) old configuration in annotations(%s) success, xx,xx,xx, OriginalSpecAnnotation.